docs: add packaging guide for contributors (#415)
ngi-nix/ngi-nix-forge: docs/manuals/contributor/how_to/package_recipe.md

# How to create a package recipe

::: {important}
For the list of all available configuration options for package recipes visit
the
[Package options reference](https://ngi-nix.github.io/forge/recipe/options?p=apps&s=packages).
:::

## Step 1: Investigate the package

> 

Reviewer: Should the contributors follow the issue triaging as well as creating tasks for themselves? If so we need to mention this clearly.

Author: Thanks for review. This is more instruction related to our task management and packaging workflow. For now, I am not sure if this information belongs to this page.

- License
- Main executable name

**Language and build system**

Look for these files in the repository root:

| File                           | Build system      |
| ------------------------------ | ----------------- |
| `configure.ac` or `configure`  | Autotools (C/C++) |
| `CMakeLists.txt`               | CMake (C/C++)     |
| `go.mod`                       | Go                |
| `Makefile`                     | Make (C/C++)      |
| `pyproject.toml` or `setup.py` | Python            |
| `Cargo.toml`                   | Rust              |

**Dependencies** — look in:

- `CMakeLists.txt`: `find_package()`, `pkg_check_modules()`
- `configure.ac`: `PKG_CHECK_MODULES`, `AC_CHECK_LIB`
- `pyproject.toml` / `setup.py`: `[project.dependencies]`, `install_requires`
- `README.md`, `INSTALL.md`: listed prerequisites
- CI config (`.github/workflows/`, `.gitlab-ci.yml`): packages installed before build

**Repository structure**

- Are build files in the repository root or a subdirectory?
- Are there git submodules (`.gitmodules` file)?
- Does the build download anything at build time? Nix builds run without network
  access - these must be patched out or disabled via build flags.

> ## Step 2: Write the recipe

Reviewer 2: Doesn't it make more sense to move the `git add` instructions here? With this, we can move the instructions to update the `source` in the Source section.

Author: Moved . Good point. I was thinking about this as well.

`packages.check` options.

```nix
  build.standardBuilder = {
    enable = true;

    # build-time dependencies
    packages.build = [
      pkgs.bison
      pkgs.boost
    ];

    # run-time dependencies
    packages.run = [
      pkgs.zlib
    ];
  }
```

Add the recipe file to Git:

```bash
git add recipes/packages/<package-name>/recipe.nix
```

::: {important}
Nix can only see recipe files tracked by Git. If the file is not added
to Git, the package will not be recognized.
:::

Now it is time to launch the first build of the package:

```bash
nix build .#<package-name> --print-build-logs
```

Nix will fail during the first build due to a missing source hash. Update
`source.hash` with the value from the error output, then launch the build once
again.

#### Troubleshooting

To troubleshoot build failures, enable _debug_ mode to launch interactive build
environment:

```nix
build.debug = true;
```

Then launch the interactive build and follow the on-screen instructions to run
each build phase manually. Press Ctrl-C to drop into a shell and start
investigating.

```bash
mkdir dev && cd dev
nix develop .#<package-name>
```

### Tests

> Add at least simple test script to verify that package works correctly:

Reviewer 2: ```suggestion
Add a test script to verify that package works correctly:
```
